File: go/jsonStorage.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"sync"
	"time"
)

var storageMutex sync.RWMutex

const (
	paymentMethodsFile = "data/payment_methods.json"
	dataDirectory     = "data"
)

// Initialize storage directory and files
func initializeStorage() error {
	// Create data directory if it doesn't exist
	if err := os.MkdirAll(dataDirectory, 0755); err != nil {
		return fmt.Errorf("failed to create data directory: %v", err)
	}

	// Initialize payment methods file if it doesn't exist
	if _, err := os.Stat(paymentMethodsFile); os.IsNotExist(err) {
		initialData := PaymentMethodsStorage{
			PaymentMethods: []PaymentMethod{},
			LastUpdated:    getCurrentTimestamp(),
		}

		data, err := json.MarshalIndent(initialData, "", "  ")
		if err != nil {
			return fmt.Errorf("failed to marshal initial data: %v", err)
		}

		if err := ioutil.WriteFile(paymentMethodsFile, data, 0644); err != nil {
			return fmt.Errorf("failed to write initial payment methods file: %v", err)
		}
	}

	return nil
}

// Load payment methods from JSON file
func loadPaymentMethods() ([]PaymentMethod, error) {
	storageMutex.RLock()
	defer storageMutex.RUnlock()

	// Check if file exists
	if _, err := os.Stat(paymentMethodsFile); os.IsNotExist(err) {
		return []PaymentMethod{}, nil
	}

	// Read file
	data, err := ioutil.ReadFile(paymentMethodsFile)
	if err != nil {
		return nil, fmt.Errorf("failed to read payment methods file: %v", err)
	}

	// Parse JSON
	var storage PaymentMethodsStorage
	if err := json.Unmarshal(data, &storage); err != nil {
		return nil, fmt.Errorf("failed to parse payment methods JSON: %v", err)
	}

	return storage.PaymentMethods, nil
}

// Save a new payment method
func savePaymentMethod(method *PaymentMethod) error {
	storageMutex.Lock()
	defer storageMutex.Unlock()

	// Load existing methods
	methods, err := loadPaymentMethodsUnsafe()
	if err != nil {
		return err
	}

	// Add new method
	methods = append(methods, *method)

	// Save back to file
	return savePaymentMethodsUnsafe(methods)
}

// Update an existing payment method
func updatePaymentMethodInStorage(updatedMethod *PaymentMethod) error {
	storageMutex.Lock()
	defer storageMutex.Unlock()

	// Load existing methods
	methods, err := loadPaymentMethodsUnsafe()
	if err != nil {
		return err
	}

	// Find and update the method
	found := false
	for i := range methods {
		if methods[i].ID == updatedMethod.ID {
			methods[i] = *updatedMethod
			found = true
			break
		}
	}

	if !found {
		return fmt.Errorf("payment method not found: %s", updatedMethod.ID)
	}

	// Save back to file
	return savePaymentMethodsUnsafe(methods)
}

// Clear default flag from all payment methods
func clearDefaultPaymentMethods() error {
	storageMutex.Lock()
	defer storageMutex.Unlock()

	// Load existing methods
	methods, err := loadPaymentMethodsUnsafe()
	if err != nil {
		return err
	}

	// Clear default flag from all methods
	for i := range methods {
		if methods[i].IsDefault {
			methods[i].IsDefault = false
			methods[i].UpdatedAt = time.Now()
		}
	}

	// Save back to file
	return savePaymentMethodsUnsafe(methods)
}

// Load payment methods without mutex (for internal use)
func loadPaymentMethodsUnsafe() ([]PaymentMethod, error) {
	// Check if file exists
	if _, err := os.Stat(paymentMethodsFile); os.IsNotExist(err) {
		return []PaymentMethod{}, nil
	}

	// Read file
	data, err := ioutil.ReadFile(paymentMethodsFile)
	if err != nil {
		return nil, fmt.Errorf("failed to read payment methods file: %v", err)
	}

	// Parse JSON
	var storage PaymentMethodsStorage
	if err := json.Unmarshal(data, &storage); err != nil {
		return nil, fmt.Errorf("failed to parse payment methods JSON: %v", err)
	}

	return storage.PaymentMethods, nil
}

// Save payment methods without mutex (for internal use)
func savePaymentMethodsUnsafe(methods []PaymentMethod) error {
	storage := PaymentMethodsStorage{
		PaymentMethods: methods,
		LastUpdated:    getCurrentTimestamp(),
	}

	// Marshal to JSON with indentation
	data, err := json.MarshalIndent(storage, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal payment methods: %v", err)
	}

	// Write to file
	if err := ioutil.WriteFile(paymentMethodsFile, data, 0644); err != nil {
		return fmt.Errorf("failed to write payment methods file: %v", err)
	}

	return nil
}

// Mock mode configuration storage
const mockModeConfigFile = "data/mock_mode_config.json"

type MockModeStorage struct {
	IsEnabled   bool   `json:"isEnabled"`
	LastUpdated string `json:"lastUpdated"`
}

// Load mock mode configuration
func loadMockModeConfig() (bool, error) {
	if _, err := os.Stat(mockModeConfigFile); os.IsNotExist(err) {
		return false, nil // Default to disabled
	}

	data, err := ioutil.ReadFile(mockModeConfigFile)
	if err != nil {
		return false, fmt.Errorf("failed to read mock mode config: %v", err)
	}

	var config MockModeStorage
	if err := json.Unmarshal(data, &config); err != nil {
		return false, fmt.Errorf("failed to parse mock mode config: %v", err)
	}

	return config.IsEnabled, nil
}

// Save mock mode configuration
func saveMockModeConfig(enabled bool) error {
	config := MockModeStorage{
		IsEnabled:   enabled,
		LastUpdated: getCurrentTimestamp(),
	}

	data, err := json.MarshalIndent(config, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal mock mode config: %v", err)
	}

	if err := ioutil.WriteFile(mockModeConfigFile, data, 0644); err != nil {
		return fmt.Errorf("failed to write mock mode config: %v", err)
	}

	return nil
}

// Delete a payment method (for future use)
func deletePaymentMethod(methodID string) error {
	storageMutex.Lock()
	defer storageMutex.Unlock()

	// Load existing methods
	methods, err := loadPaymentMethodsUnsafe()
	if err != nil {
		return err
	}

	// Find and remove the method
	found := false
	for i := range methods {
		if methods[i].ID == methodID {
			// Remove element at index i
			methods = append(methods[:i], methods[i+1:]...)
			found = true
			break
		}
	}

	if !found {
		return fmt.Errorf("payment method not found: %s", methodID)
	}

	// Save back to file
	return savePaymentMethodsUnsafe(methods)
}

File: nodejs/jsonStorage.js
/**
 * JSON Storage utility for payment methods and configuration
 * Node.js implementation with async file operations
 */

import fs from 'fs/promises';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';

const DATA_DIR = 'data';
const PAYMENT_METHODS_FILE = path.join(DATA_DIR, 'payment_methods.json');
const MOCK_MODE_CONFIG_FILE = path.join(DATA_DIR, 'mock_mode_config.json');

/**
 * Initialize storage directory and files
 */
export async function initializeStorage() {
    try {
        // Create data directory if it doesn't exist
        await fs.mkdir(DATA_DIR, { recursive: true });
        
        // Initialize payment methods file if it doesn't exist
        try {
            await fs.access(PAYMENT_METHODS_FILE);
        } catch {
            await fs.writeFile(PAYMENT_METHODS_FILE, '[]', 'utf8');
        }
        
        console.log('✅ Storage initialized successfully');
    } catch (error) {
        console.error('❌ Failed to initialize storage:', error.message);
        throw error;
    }
}

/**
 * Generate unique ID for payment methods
 */
function generateId() {
    return `pm_${Date.now()}_${uuidv4().replace(/-/g, '').substring(0, 16)}`;
}

/**
 * Load all payment methods
 */
export async function loadPaymentMethods() {
    try {
        const data = await fs.readFile(PAYMENT_METHODS_FILE, 'utf8');
        return JSON.parse(data);
    } catch (error) {
        console.log('Warning: Could not load payment methods, returning empty array');
        return [];
    }
}

/**
 * Save payment methods to file
 */
async function savePaymentMethods(methods) {
    try {
        await fs.writeFile(PAYMENT_METHODS_FILE, JSON.stringify(methods, null, 2), 'utf8');
    } catch (error) {
        console.error('Error saving payment methods:', error.message);
        throw error;
    }
}

/**
 * Add a new payment method
 */
export async function addPaymentMethod(data) {
    const methods = await loadPaymentMethods();
    
    // Generate unique ID
    const id = generateId();
    
    // Create payment method object
    const paymentMethod = {
        id: id,
        storedPaymentToken: data.storedPaymentToken,
        cardBrand: data.cardBrand,
        last4: data.last4,
        expiry: data.expiry,
        nickname: data.nickname || `${data.cardBrand} ending in ${data.last4}`,
        isDefault: data.isDefault || false,
        createdAt: new Date().toISOString(),
        updatedAt: new Date().toISOString()
    };
    
    // Handle default payment method
    if (paymentMethod.isDefault) {
        // Remove default from all existing methods
        methods.forEach(method => {
            method.isDefault = false;
        });
    } else if (methods.length === 0) {
        // Make first method default
        paymentMethod.isDefault = true;
    }
    
    methods.push(paymentMethod);
    await savePaymentMethods(methods);
    
    return paymentMethod;
}

/**
 * Find payment method by ID
 */
export async function findPaymentMethod(id) {
    const methods = await loadPaymentMethods();
    return methods.find(method => method.id === id) || null;
}

/**
 * Update an existing payment method (nickname and default status only)
 */
export async function updatePaymentMethod(id, updateData) {
    const methods = await loadPaymentMethods();
    const methodIndex = methods.findIndex(method => method.id === id);
    
    if (methodIndex === -1) {
        return null;
    }
    
    const method = methods[methodIndex];
    
    // Update editable fields
    if (updateData.nickname !== undefined) {
        method.nickname = updateData.nickname;
    }
    
    // Handle default payment method change
    if (updateData.isDefault !== undefined) {
        if (updateData.isDefault && !method.isDefault) {
            // Remove default from all existing methods
            methods.forEach(m => {
                m.isDefault = false;
            });
            method.isDefault = true;
        } else if (!updateData.isDefault && method.isDefault) {
            method.isDefault = false;
        }
    }
    
    method.updatedAt = new Date().toISOString();
    await savePaymentMethods(methods);
    
    return method;
}

/**
 * Get all payment methods formatted for display
 */
export async function getFormattedPaymentMethods() {
    const methods = await loadPaymentMethods();
    
    return methods.map(method => ({
        id: method.id,
        type: 'card',
        last4: method.last4,
        brand: method.cardBrand,
        expiry: method.expiry,
        nickname: method.nickname || '',
        isDefault: method.isDefault || false
    }));
}

/**
 * Set a payment method as default (and remove default from others)
 */
export async function setDefaultPaymentMethod(id) {
    const methods = await loadPaymentMethods();
    const targetMethod = methods.find(method => method.id === id);
    
    if (!targetMethod) {
        return false;
    }
    
    // Remove default from all methods
    methods.forEach(method => {
        method.isDefault = false;
    });
    
    // Set target method as default
    targetMethod.isDefault = true;
    targetMethod.updatedAt = new Date().toISOString();
    
    await savePaymentMethods(methods);
    return true;
}

/**
 * Delete a payment method (for future use)
 */
export async function deletePaymentMethod(id) {
    const methods = await loadPaymentMethods();
    const filteredMethods = methods.filter(method => method.id !== id);
    
    if (filteredMethods.length === methods.length) {
        return false; // Method not found
    }
    
    await savePaymentMethods(filteredMethods);
    return true;
}

/**
 * Load mock mode configuration
 */
export async function loadMockModeConfig() {
    try {
        const data = await fs.readFile(MOCK_MODE_CONFIG_FILE, 'utf8');
        const config = JSON.parse(data);
        return config.isEnabled || false;
    } catch {
        return false; // Default to disabled
    }
}

/**
 * Save mock mode configuration
 */
export async function saveMockModeConfig(enabled) {
    const config = {
        isEnabled: enabled,
        lastUpdated: new Date().toISOString()
    };
    
    try {
        await fs.writeFile(MOCK_MODE_CONFIG_FILE, JSON.stringify(config, null, 2), 'utf8');
    } catch (error) {
        console.error('Error saving mock mode config:', error.message);
        throw error;
    }
}